File: tests/basic/typedef_fold.ml
open Optitrust
open Target


let _ = Run.doc_script_cpp (fun _ ->
    !! Typedef_basic.fold ~at:[cVarDef "c"] [cTypDef "uchar"];
  )
"
typedef unsigned char uchar;

unsigned char c;
"

(* LATER: figure out why if our printing of unsigned/long follows the standard recommendations *)

let _ = Run.script_cpp (fun _ ->
  !! Typedef_basic.fold ~at:[cTypDef "vect"] [cTypDef "uint"];
  !! Typedef_basic.fold ~at:[cVarDef "y1"] [cTypDef "cdouble"];
  !! Typedef_basic.fold [cTypDef "mat2d"] ;
  !! Typedef_basic.fold [cTypDef "mat3d"];
  )

File: tests/basic/typedef_fold.cpp
typedef unsigned int uint;

typedef double const cdouble;

typedef struct { unsigned int x; unsigned int y; } vect;

typedef int** mat2d;

typedef int*** mat3d; // to be folded before mat2d, else M would be of type mat2d*

int main() {
  unsigned int x;
  unsigned int t[3];
  int a = 1;
  int b;
  const double y1 = 1.0;
  double const y2 = 2.0;
  vect v;
  int** m;
  int*** M;
  int**** T;
}


